### ISSUE TITLE
Struct getters are generating unneccessary instructions on x64

### ISSUE THREAD

**Issue by GeorchW** [dotnet/coreclr#8828]
Hi folks,

I hope I'm writing this in the right area. I was comparing the JIT outputs of properties vs. fields, to see whether I should avoid auto-properties in high-performance code. Now, I've managed to create an example where the exact same data embedded in a struct creates different output than just using a primitive type. To be exact, it looks like the entire struct is copied once more than neccessary. Additionally, the resulting code is significantly larger when embedding floats as opposed to embedding ints, even though the data is only copied around.

The behaviour when using primitive types seems to be as expected (no additional instructions when using properties).

I'm using VS2017 RC, .NET 4.6.2, compiling to x64, I think it should be RyuJIT, although I don't know how to find that out. 

Example code:

Disassembly:
With struct { public float Value; }

             testClass.Field = testStruct;
     mov         qword ptr [rsp+10h],rdx  
     mov         eax,dword ptr [rsp+10h]  
     mov         dword ptr [rcx+8],eax  
     ret  
     
             testClass.Property = testStruct;
     sub         rsp,8  
     mov         qword ptr [rsp+18h],rdx  
     mov         eax,dword ptr [rsp+18h]  
     mov         dword ptr [rsp],eax  
     mov         eax,dword ptr [rsp]  
     mov         dword ptr [rcx+10h],eax  
     add         rsp,8  
     ret  
     
             Print(testClass.Field);
     sub         rsp,28h  
     mov         ecx,dword ptr [rcx+8]  
     call        00007FFC579A00A0  
     nop  
     add         rsp,28h  
     ret  
     
             Print(testClass.Property);
     sub         rsp,28h  
     mov         ecx,dword ptr [rcx+10h]  
     mov         dword ptr [rsp+20h],ecx  
     mov         ecx,dword ptr [rsp+20h]  
     call        00007FFC579A05A0  
     nop  
     add         rsp,28h  
     ret  
     
With struct { public int Value; }

            testClass.Field = testStruct;
     mov         dword ptr [rcx+8],edx  
     ret  
     
             testClass.Property = testStruct;
     mov         eax,edx  
     mov         dword ptr [rcx+10h],eax  
     ret  
     
             Print(testClass.Field);
     sub         rsp,28h  
     mov         ecx,dword ptr [rcx+8]  
     call        00007FFC579C00A0  
     nop  
     add         rsp,28h  
     ret  
     
             Print(testClass.Property);
     sub         rsp,28h  
     mov         ecx,dword ptr [rcx+10h]  
     call        00007FFC579C0560  
     nop  
     add         rsp,28h  
     ret  

Full source:

    using System;
    using System.Runtime.CompilerServices;

    struct TestStruct
    {
        public float Value;
        //public int Value;
    }
    class TestClass
    {
        public TestStruct Field;
        public TestStruct Property { get; set; }
    }
    class Program
    {

        static void Main(string[] args)
        {
            TestClass testClass = new TestClass();
            TestFieldSetters(testClass, new TestStruct() { Value = 5 });
            TestPropSetters(testClass, new TestStruct() { Value = 3 });
            TestFieldGetters(testClass);
            TestPropGetters(testClass);
            Console.ReadLine();
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void TestFieldSetters(TestClass testClass, TestStruct testStruct)
        {
            testClass.Field = testStruct;
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void TestPropSetters(TestClass testClass, TestStruct testStruct)
        {
            testClass.Property = testStruct;
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void TestFieldGetters(TestClass testClass)
        {
            Print(testClass.Field);
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void TestPropGetters(TestClass testClass)
        {
            Print(testClass.Property);
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void Print(TestStruct testStruct)
        {
            Console.WriteLine(testStruct.Value);
        }
    }

**Comment by mikedn** [comment]
> To be exact, it looks like the entire struct is copied once more than neccessary. 

Yep, accessing a field via a property naturally results in a temporary copy and the JIT can't always eliminate that. See #1161 as well.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib
